### DISCUSSION TITLE
결제 완료시 포인트 적립 처리 방법을 어떻게 해야될까요?

### DISCUSSION THREAD

**Discussion by f1v3-dev** [nhnacademy-be4-ckin/ckin-api#133] [Ideas]
> ### 함께 고민해보고싶은 주제가 있습니다. (혼자서는 어떻게 처리해야될지 모르겠어요.. 😂)

결제가 완료되면 Front 서버에서 API 서버로 결제 정보를 넘겨주고, 
넘겨받은 데이터를 DB **결제(Payment) 테이블에 저장**하게됩니다.

그런데, 결제가 완료된 시점에 `주문한 금액 * 회원 등급 포인트 정책` 만큼 **포인트 적립 로직도 실행**합니다.
또한, 포인트가 적립되었으므로 포인트 내역에 `주문 적립` 이라는 사유로 **내역 등록** 진행해야합니다.

이러한 경우에 결제완료와 포인트 적립( + 내역 등록)을 한 개의 트랜잭션으로 묶어버리게 된다면,
결제가 완료되었는데 포인트 적립에서 문제가 발생해도 결제 완료 처리가 되지 않습니다.

따라서 결제 컨트롤러에서

```java
    @PostMapping
    public ResponseEntity<PaymentSuccessResponseDto> createPayment(@RequestBody PaymentRequestDto requestDto) {
        // 결제 SAVE
        PaymentSuccessResponseDto payment = paymentFacade.createPayment(requestDto);

        // 포인트 적립 + 포인트 기록
        paymentFacade.createRewardPoint(payment.getSaleNumber());
        return ResponseEntity.status(HttpStatus.CREATED).body(payment);
    }
```

위와 같이 호출하고, Service(Facade)

```java
    @Transactional
    public PaymentSuccessResponseDto createPayment(PaymentRequestDto requestDto) {

        ...

        paymentService.createPayment(sale.getSaleId(), requestDto);
        saleService.updateSalePaymentPaidStatus(sale.getSaleId());

        return PaymentSuccessResponseDto.builder()
                .saleNumber(sale.getSaleNumber())
                ...
                .build();
    }

    @Transactional
    public void createRewardPoint(String saleNumber) {
        SaleResponseDto sale = saleService.getSaleBySaleNumber(saleNumber);

        if (Objects.nonNull(sale.getMemberEmail())) {
            memberService.updateRewardPoint(sale.getMemberEmail(), sale.getSaleTotalPrice());
        }
    }
```

다음과 같이 개별 트랜잭션으로 진행하도록 하였는데, 이게 맞는 과정인지 궁금합니다. 

다른 분들의 의견이 궁금해요!

**Comment by devhomh** [comment]
궁금한게 있는데 결제가 완료된 상태에서 
` paymentFacade.createPayment(requestDto);`
이 부분에서 에러가 나면 결제가 다시 취소가 되나요?

**Reply by f1v3-dev** [reply]
토스 페이먼츠쪽에서 결제 완료 상태로 변경이 되어서 결제가 취소되지는 않을 것 같아요
대신, DB상에선 결제가 완료되지 않은 주문이라고 뜨는 문제가 있긴 하네요 😅 
데이터 정합성 문제에 대해서 조금 더 고민해볼게요 감사합니다 !

**Comment by dduneon** [comment]
좋은거 같아요! 저라도 그렇게 했을것 같네요
결제는 완료되었는데 포인트 적립이 실패했을 때에는 로그 찍어서 수동 적립시킬 수 있도록 그런 방안도 생각해보면 좋을것 같슴당

**Comment by f1v3-dev** [comment]
## 해결 방안

> @unhas01 경서 TA님의 도움으로 `@Transactional - propagation` 옵션에 대해서 살펴보았습니다. 😊

### 트랜잭션 전파 속성
![image](https://github.com/nhnacademy-be4-ckin/ckin-api/assets/84575041/1cc984bd-a910-496c-92b4-c99bfffa3de9)

JPA에서 제공하는 `@Transactioanl` 어노테이션의 _default propagation_ 옵션은 `REQUIRED` 입니다.
이 옵션의 경우 기존 트랜잭션에 포함된다는 특징을 가지고 있습니다.

이러한 경우에 본문의 코드에서는 **기대했던 결과와 다르게** 개별 트랜잭션으로 나눠지지 않고 **기존 트랜잭션**에 참여하게 된다고 합니다.

따라서 `memberService.updateRewardPoint()` 에서 `propagation` 옵션을 `REQUIRES_NEW`로 설정해주는 작업을 진행하였습니다.
또한, `paymentFacade.createPayment()` 하나의 메서드에서 공통 처리로 변경하였습니다.

```java

    @Override
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void updateRewardPoint(String email, Integer totalPrice) {

       ...
        member.updatePoint(reward);

        PointHistory pointHistory = PointHistory.builder()
                .member(member)
                .pointHistoryPoint(reward)
                .pointHistoryReason("주문 적립")
                .pointHistoryTime(LocalDate.now())
                .build();

        pointHistoryRepository.save(pointHistory);
    }
```

 > 이해하기 어려울 수 있어, 아래의 사진도 첨부합니다 !
 
- `물리 트랜잭션`: 실제 데이터베이스에 적용되는 트랜잭션으로, 커넥션을 통해 커밋/롤백하는 단위
- `논리 트랜잭션`: 스프링이 트랜잭션 매니저를 통해 트랜잭션을 처리하는 단위

 #### REQUIRED
 ![image](https://github.com/nhnacademy-be4-ckin/ckin-api/assets/84575041/58093501-dffe-4110-a677-dd01dbfeb257)


 #### REQUIRES_NEW
![image](https://github.com/nhnacademy-be4-ckin/ckin-api/assets/84575041/fd1ef391-ab0d-4293-817c-dc575fdde123)




> 참고
> [[Spring] 스프링의 트랜잭션 전파 속성(Transaction propagation) 완벽하게 이해하기](https://mangkyu.tistory.com/269)
> [[Spring] @Transactional 옵션 알아보기 + 트랜잭션 전파](https://hstory0208.tistory.com/entry/Spring-Transactional-%EC%98%B5%EC%85%98-%EC%95%8C%EC%95%84%EB%B3%B4%EA%B8%B0-%ED%8A%B8%EB%9E%9C%EC%9E%AD%EC%85%98-%EC%A0%84%ED%8C%8C)

**Comment by f1v3-dev** [comment]
### 오해하고 있었던 내용

> [[Spring] Transactional Propagation 정리하기](https://devlog-wjdrbs96.tistory.com/424)

![image](https://github.com/nhnacademy-be4-ckin/ckin-api/assets/84575041/8c330735-ed63-4649-9afa-692b7179e84d)

- Transaction 전파와 Exception 전파는 별도로 처리해줘야 합니다.

따라서, 코드 상에서도 try-catch 처리를 해주지 않을 경우 **Exception**이 상위로 올라가게 되어, 결제에 대한 처리도 `rollback` 되어버리는 문제점이 여전히 존재했었으나, try-catch 처리를 진행하였습니다.
